### ISSUE TITLE
notebook versioning via git

### ISSUE THREAD

**Issue by vidma** [andypetrella/spark-notebook#479]
Today we came up with, IMHO, a pretty good idea, to use `git` for notebook versioning.

- [ ] `conf/application.conf` option to use `*.ipynb` extension, so it's displayed nicely on github
- [ ] save a checkpoint (optionally with a name given)

  `git add dir/notebook-name.snb && git commit -m 'Checkpoint: User given name'`
- [ ] list existing checkpoints
- [ ] restore to a certain checkpoint

  `git reset commitSHA -- filename && git commit -m 'Restore to a checkpoint'`
- [ ] git commands need to be synchronized, so it's one commit per notebook
- [ ] push to github (can be done manually at first)
- [ ] store graphs/tables in `snb` readable format
  * graphs: svg->png
  * tables

@andypetrella 
comments welcome :)

**Comment by andypetrella** [comment]
looks nice!

is it only for checkpointing?

because I was considering as well putting an abstraction in the notebook manager to allow loading and saving from several sources, this would be complimentary I guess?

**Comment by vidma** [comment]
> is it only for checkpointing?

if we integrated `git`, github would be a natural choice :)

**Comment by andypetrella** [comment]
if it's public yes, but it needs to be pluggable...

**Comment by vidma** [comment]
> if it's public yes, but it needs to be pluggable...

I see no problem with private repos either, it's just a matter of adding a public/private-key or smf like that.

**Comment by andypetrella** [comment]
yup,  I meant bit bucket or other internal git.  but support of bare git is probably what you meant, not just github? 

Also there is also s3 or swift or simply local folder I guess

**Comment by vidma** [comment]
> but support of bare git is probably what you meant, not just github?

exactly.

**Comment by andypetrella** [comment]
I'm up for that, we can suppose that git is installed and available everywhere, now we'll need to `git init`.

I'd love to see if we can come up directly with both support of local filesystem and git, let's see how it'll go

**Comment by shijinkui** [comment]
sbt-release

**Comment by andypetrella** [comment]
@shijinkui what do you mean exaclty? I'm not quite sure actually if you propose us to use sbt release for the notebooks or for the spark-notebook project :-D

**Comment by shijinkui** [comment]
@andypetrella using `sbt-release` to auto increase version number is more convenienced :)
